main.tf

```diff
@@ -1,3 +1,27 @@
+locals {
+
+  # Abstract the decision whether to create an Analytics Workspace or not.
+  create_analytics_workspace = var.log_analytics_workspace_enabled && var.log_analytics_workspace == null
+
+  # Abstract the decision whether to use an Analytics Workspace supplied via vars, provision one ourselves or leave it null.
+  # This guarantees that local.log_analytics_workspace will contain a valid `id` and `name` IFF log_analytics_workspace_enabled
+  # is set to `true`.
+  log_analytics_workspace = var.log_analytics_workspace_enabled ? (
+    # The Log Analytics Workspace should be enabled:
+    var.log_analytics_workspace == null ? {
+      # `log_analytics_workspace_enabled` is `true` but `log_analytics_workspace` was not supplied.
+      # Create an `azurerm_log_analytics_workspace` resource and use that.
+      id   = azurerm_log_analytics_workspace.main[0].id
+      name = azurerm_log_analytics_workspace.main[0].name
+      } : {
+      # `log_analytics_workspace` is supplied. Let's use that.
+      id   = var.log_analytics_workspace.id
+      name = var.log_analytics_workspace.name
+    }
+  ) : null # Finally, the Log Analytics Workspace should be disabled.
```

Comment:
Would an object with `null` field be better (the caller don't need to check whether this object is `null`)?

```hcl
) : {
  id       = null
  name = null
}

Reply:
But won't the caller just have to check whether `local.log_analytics_workspace.id/name` is `null` instead? 

I feel that this breaks the Principle of Least Astonishment (maybe my background in `python` is showing).

Reply:
In Terraform assign `null` to an argument is equal to omit this argument, in some cases the caller doesn't need to check whether the log analytics workspace id is `null` or not, if the caller just assigns it to an optional argument, in that case we can save a `null` check. I personally prefer this "`null` safe" style, but I think your point also make sense. Please allow me to have more discuss with other people, thanks for your understanding.

Reply:
did we reach a consensus on:
```
) : {
  id       = null
  name = null
}
```
vs. 
```
) : null
```
?